File: android/src/main/java/com/lebrislo/bluetooth/mesh/plugin/SigPluginCall.kt
package com.lebrislo.bluetooth.mesh.plugin

import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import no.nordicsemi.android.mesh.transport.GenericLevelStatus
import no.nordicsemi.android.mesh.transport.GenericOnOffStatus
import no.nordicsemi.android.mesh.transport.GenericPowerLevelStatus
import no.nordicsemi.android.mesh.transport.LightCtlStatus
import no.nordicsemi.android.mesh.transport.LightCtlTemperatureRangeStatus
import no.nordicsemi.android.mesh.transport.LightHslStatus
import no.nordicsemi.android.mesh.transport.MeshMessage

/**
 * This class is used to generate a response for a SIG plugin call.
 */
class SigPluginCall(val meshOperationCallback: Int, val meshAddress: Int, call: PluginCall) : BasePluginCall(call) {
    companion object {
        private val tag: String = SigPluginCall::class.java.simpleName

        /**
         * Generates a response for a SIG plugin call.
         *
         * @param meshMessage Mesh message.
         */
        @JvmStatic
        fun generateSigPluginCallResponse(meshMessage: MeshMessage): JSObject {
            val result = JSObject()
            result.put("src", meshMessage.src)
            result.put("dst", meshMessage.dst)
            result.put("opcode", meshMessage.opCode)
            result.put(
                "data", when (meshMessage) {
                    is GenericOnOffStatus -> genericOnOffStatusResponse(meshMessage)
                    is GenericLevelStatus -> genericLevelStatusResponse(meshMessage)
                    is GenericPowerLevelStatus -> genericPowerLevelStatusResponse(meshMessage)
                    is LightHslStatus -> lightHslStatusResponse(meshMessage)
                    is LightCtlStatus -> lightCtlStatusResponse(meshMessage)
                    is LightCtlTemperatureRangeStatus -> lightCtlTemperatureRangeStatusResponse(meshMessage)
                    else -> JSObject()
                }
            )
            Log.d(tag, "generateSigPluginCallResponse: $result")
            return result
        }

        private fun genericOnOffStatusResponse(meshMessage: GenericOnOffStatus): JSObject {
            val data = JSObject()
            data.put("onOff", meshMessage.parameters[0].toInt() == 1)
            return data
        }

        private fun genericLevelStatusResponse(meshMessage: GenericLevelStatus): JSObject {
            val data = JSObject()
            data.put("level", meshMessage.presentLevel)
            return data
        }

        private fun genericPowerLevelStatusResponse(meshMessage: GenericPowerLevelStatus): JSObject {
            val data = JSObject()
            data.put("powerLevel", meshMessage.presentLevel.toUShort().toInt())
            return data
        }

        private fun lightHslStatusResponse(meshMessage: LightHslStatus): JSObject {
            val data = JSObject()
            data.put("hue", meshMessage.presentHue.toUShort().toInt())
            data.put("saturation", meshMessage.presentSaturation.toUShort().toInt())
            data.put("lightness", meshMessage.presentLightness.toUShort().toInt())
            return data
        }

        private fun lightCtlStatusResponse(meshMessage: LightCtlStatus): JSObject {
            val data = JSObject()
            data.put("lightness", meshMessage.presentLightness.toUShort().toInt())
            data.put("temperature", meshMessage.presentTemperature.toUShort().toInt())
            return data
        }

        private fun lightCtlTemperatureRangeStatusResponse(meshMessage: LightCtlTemperatureRangeStatus): JSObject {
            val data = JSObject()
            data.put("status", meshMessage.statusCode.toUShort().toInt())
            data.put("min", meshMessage.rangeMin.toUShort().toInt())
            data.put("max", meshMessage.rangeMax.toUShort().toInt())
            return data
        }
    }
}

File: ios/Sources/BluetoothMeshPlugin/plugin/SigPluginCall.swift
//
//  SigPluginCall.swift
//  BluetoothMeshPlugin
//
//  Created by LE BRIS Loris on 14/11/2025.
//

import Capacitor
import Foundation
import NordicMesh

public class SigPluginCall: BasePluginCall {
    public let meshOperationCallback: UInt32
    public let meshAddress: UInt16

    public init(_ meshOperationCallback: UInt32, _ meshAddress: UInt16, _ call: CAPPluginCall) {
        self.meshOperationCallback = meshOperationCallback
        self.meshAddress = meshAddress
        super.init(call: call)
    }

    public static func generateSigPluginCallResponse(_ response: RoutedMeshMessage) -> PluginCallResultData {
        var result: PluginCallResultData = [:]
        result["src"] = response.src
        result["dst"] = response.dst
        result["opcode"] = response.message.opCode

        if let msg = response.message as? GenericOnOffStatus {
            result["data"] = genericOnOffStatusResponse(msg)
        } else if let msg = response.message as? GenericLevelStatus {
            result["data"] = genericLevelStatusResponse(msg)
        } else if let msg = response.message as? GenericPowerLevelStatus {
            result["data"] = genericPowerLevelStatusResponse(msg)
        } else if let msg = response.message as? LightHSLStatus {
            result["data"] = lightHslStatusResponse(msg)
        } else if let msg = response.message as? LightCTLStatus {
            result["data"] = lightCtlStatusResponse(msg)
        } else if let msg = response.message as? LightCTLTemperatureRangeStatus {
            result["data"] = lightCtlTemperatureRangeStatusResponse(msg)
        } else if let msg = response.message as? HealthFaultStatus {
            result["data"] = healthFaultStatusResponse(msg)
        } else if let msg = response.message as? HealthCurrentStatus {
            result["data"] = healthCurrentStatusResponse(msg)
        } else {
            result["data"] = PluginCallResultData()
        }
        return result
    }

    private static func genericOnOffStatusResponse(_ msg: GenericOnOffStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["onOff"] = msg.isOn
        return data
    }

    private static func genericLevelStatusResponse(_ msg: GenericLevelStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["level"] = msg.level
        return data
    }

    private static func genericPowerLevelStatusResponse(_ msg: GenericPowerLevelStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["powerLevel"] = msg.power
        return data
    }

    private static func lightHslStatusResponse(_ msg: LightHSLStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["hue"] = msg.hue
        data["saturation"] = msg.saturation
        data["lightness"] = msg.lightness
        return data
    }

    private static func lightCtlStatusResponse(_ msg: LightCTLStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["lightness"] = msg.lightness
        data["temperature"] = msg.temperature
        return data
    }

    private static func lightCtlTemperatureRangeStatusResponse(_ msg: LightCTLTemperatureRangeStatus)
        -> PluginCallResultData
    {
        var data: PluginCallResultData = [:]
        data["status"] = msg.status
        data["min"] = msg.min
        data["max"] = msg.max
        return data
    }

    private static func healthFaultStatusResponse(_ msg: HealthFaultStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["testId"] = msg.testId
        data["companyId"] = msg.companyIdentifier
        data["faults"] = msg.faults
        return data
    }

    private static func healthCurrentStatusResponse(_ msg: HealthCurrentStatus) -> PluginCallResultData {
        var data: PluginCallResultData = [:]
        data["testId"] = msg.testId
        data["companyId"] = msg.companyIdentifier
        data["faults"] = msg.faults
        return data
    }
}
